cirq/qis/states.py

```diff
@@ -321,6 +313,231 @@ def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
     return (dim,)
 
 
+def infer_qid_shape(*states: 'cirq.QUANTUM_STATE_LIKE') -> Tuple[int, ...]:
+    """Infer the qid shape of a set of states.
+
+    This is a heuristic that is guaranteed to return a qid shape compatible
```

Comment:
I think we have different definitions of guarantee. 60% of the time it works every time